### ISSUE TITLE
Monitor recap alert sweep index

### ISSUE THREAD

**Issue by mlissner** [freelawproject/courtlistener#4694]
We need to monitor the new sweep index. From our earlier PR:

> Do we just need to check on the cronjob tomorrow, then, I assume?

Yes, we could also re-check the ES query to confirm that the reindexing task is running.

`GET _tasks?detailed=true&actions=*reindex`

_Originally posted by @albertisfu in https://github.com/freelawproject/courtlistener/issues/4672#issuecomment-2477668050_

**Comment by mlissner** [comment]
@albertisfu, I spun this into a new issue, so we can discuss it here. I ran the tasks command just now and it actually had results:

```
{
  "nodes": {
    "nL1wCXyqQCW64ODQvubEnA": {
      "name": "elastic-cluster-es-master-data-nodes-v3-5",
      "transport_address": "172.30.0.203:9300",
      "host": "172.30.0.203",
      "ip": "172.30.0.203:9300",
      "roles": [
        "data",
        "data_cold",
        "data_content",
        "data_frozen",
        "data_hot",
        "data_warm",
        "ingest",
        "master",
        "ml",
        "remote_cluster_client",
        "transform"
      ],
      "attributes": {
        "ml.allocated_processors": "4",
        "k8s_node_name": "ip-172-30-0-75.us-west-2.compute.internal",
        "ml.allocated_processors_double": "4.0",
        "ml.machine_memory": "31138512896",
        "xpack.installed": "true",
        "ml.max_jvm_size": "17179869184"
      },
      "tasks": {
        "nL1wCXyqQCW64ODQvubEnA:1261302059": {
          "node": "nL1wCXyqQCW64ODQvubEnA",
          "id": 1261302059,
          "type": "transport",
          "action": "indices:data/write/reindex",
          "status": {
            "total": 792895,
            "updated": 0,
            "created": 390000,
            "deleted": 0,
            "batches": 391,
            "version_conflicts": 0,
            "noops": 0,
            "retries": {
              "bulk": 0,
              "search": 0
            },
            "throttled_millis": 0,
            "requests_per_second": -1,
            "throttled_until_millis": 0
          },
          "description": "reindex from [recap_vectors] to [recap_sweep]",
          "start_time_in_millis": 1731628870357,
          "running_time_in_nanos": 2474749945113,
          "cancellable": true,
          "cancelled": false,
          "headers": {}
        }
      }
    }
  }
}
```

Surprising, no?

**Comment by mlissner** [comment]
Well, this took me down a rabbit hole, but luckily I had some time. We didn't have timezone information set on this or any other of our cronjobs, so I just audited and tweaked them all. This means this won't run until tonight, so we'll have to check on it on Monday.

I did clear out the redis stuff a second time though, so that should be good.

**Comment by ERosendo** [comment]
Here are the logs from the latest execution on November 15th:

```logs
2024-11-15 20:01:09.712 INFO Re-indexing task scheduled ID: 0cQl85qiTiyiNppgQkkPOA:1952097644
2024-11-15 20:02:09.717 INFO Task progress: 26000/627645 documents. Estimated time to finish: 1388.953806 seconds.
2024-11-15 20:17:09.720 INFO Task progress: 26000/627645 documents. Estimated time to finish: 22215.187001 seconds.
2024-11-15 20:32:09.723 INFO Task progress: 192000/627645 documents. Estimated time to finish: 4220.377473 seconds.
2024-11-15 20:47:09.727 INFO Task progress: 315000/627645 documents. Estimated time to finish: 2739.398105 seconds.
2024-11-15 21:02:09.730 INFO Task progress: 402000/627645 documents. Estimated time to finish: 2054.400233 seconds.
2024-11-15 21:17:09.734 INFO Task progress: 507000/627645 documents. Estimated time to finish: 1085.100631 seconds.
2024-11-15 21:32:09.738 INFO Task progress: 625000/627645 documents. Estimated time to finish: 60.0 seconds.

2024-11-15 21:33:09.760 INFO Resuming re-indexing process for date: 2024-11-15 00:00:00
2024-11-15 21:33:10.286 INFO Re-indexing task scheduled ID: 0cQl85qiTiyiNppgQkkPOA:1952406648
2024-11-15 21:34:10.289 INFO Task progress: 9000/564072 documents. Estimated time to finish: 3701.636893 seconds.
2024-11-15 21:49:10.292 INFO Task progress: 9000/564072 documents. Estimated time to finish: 59208.99034 seconds.
2024-11-15 22:04:10.296 INFO Task progress: 150000/564072 documents. Estimated time to finish: 5134.561793 seconds.
2024-11-15 22:19:10.299 INFO Task progress: 283000/564072 documents. Estimated time to finish: 2741.225053 seconds.
2024-11-15 22:34:10.303 INFO Task progress: 395000/564072 documents. Estimated time to finish: 1566.604886 seconds.
2024-11-15 22:49:10.306 INFO Task progress: 537000/564072 documents. Estimated time to finish: 229.886932 seconds.

```

I've included timestamps to help us understand how long it takes to reindex records.

It looks like the  [missing_error_index issue](https://github.com/freelawproject/courtlistener/issues/4646) is resolved. However, the command was unable to complete due to issue #4698.

**Comment by mlissner** [comment]
Cool. Sounds like we should close #4646 and put #4698 on our next sprint, with this as a parent?

**Comment by ERosendo** [comment]
> Sounds like we should close https://github.com/freelawproject/courtlistener/issues/4646 and put https://github.com/freelawproject/courtlistener/issues/4698 on our next sprint, with this as a parent?

Sounds good! Let's close #4646 and move #4698 to the next sprint.

**Comment by mlissner** [comment]
I want to get this off my backlog, because (1), I'm not making progress on it, and (2), I don't want to be in the way. We're not currently working on alerts, so when we get back to alerts, let's figure out what's needed here and then we can revisit it properly.

**Comment by albertisfu** [comment]
After the tweaks applied in https://github.com/freelawproject/courtlistener/pull/5316 the `cl_send_recap_alerts` command has been properly running every night.

Here the logs from last night:

```
2025-04-10 04:19:56.936 INFO Task progress: 1455000/1497073 documents. Estimated time to finish: 152.316363 seconds.
2025-04-10 04:08:09.448 INFO Task progress: 1296000/1497073 documents. Estimated time to finish: 707.483733 seconds.
2025-04-10 03:53:09.444 INFO Task progress: 1076000/1497073 documents. Estimated time to finish: 1432.284492 seconds.
2025-04-10 03:38:09.440 INFO Task progress: 671000/1497073 documents. Estimated time to finish: 3397.883109 seconds.
2025-04-10 03:23:09.436 INFO Task progress: 295000/1497073 documents. Estimated time to finish: 7579.244708 seconds.
2025-04-10 03:08:09.431 INFO Task progress: 25000/1497073 documents. Estimated time to finish: 56528.40442 seconds.
2025-04-10 02:53:09.425 INFO Task progress: 25000/1497073 documents. Estimated time to finish: 3533.421297 seconds.
2025-04-10 02:52:09.421 INFO Re-indexing task for index recap_document_sweep scheduled with ID: 5ONTTNn-S9ac4XievrZRLw:263997671
2025-04-10 02:52:08.892 INFO Deleting recap_document_sweep index from previous day.
2025-04-10 02:52:08.891 INFO Starting recap_document_sweep re-indexing process for date: 2025-04-09 00:00:00


2025-04-10 02:46:42.580 INFO Task progress: 1554000/1634077 documents. Estimated time to finish: 326.303597 seconds.
2025-04-10 02:32:10.361 INFO Task progress: 1409000/1634077 documents. Estimated time to finish: 872.214705 seconds.
2025-04-10 02:17:10.356 INFO Task progress: 1227000/1634077 documents. Estimated time to finish: 1512.895853 seconds.
2025-04-10 02:02:10.350 INFO Task progress: 913000/1634077 documents. Estimated time to finish: 2890.723157 seconds.
2025-04-10 01:47:10.346 INFO Task progress: 563000/1634077 documents. Estimated time to finish: 5250.975071 seconds.
2025-04-10 01:32:10.244 INFO Task progress: 314000/1634077 documents. Estimated time to finish: 7819.632358 seconds.
2025-04-10 01:17:10.238 INFO Task progress: 34000/1634077 documents. Estimated time to finish: 45179.14873 seconds.
2025-04-10 01:02:10.235 INFO Task progress: 34000/1634077 documents. Estimated time to finish: 2824.017311 seconds.
2025-04-10 01:01:10.230 INFO Re-indexing task for index recap_sweep scheduled with ID: 5ONTTNn-S9ac4XievrZRLw:263692823
2025-04-10 01:01:09.423 INFO Deleting recap_sweep index from previous day.
2025-04-10 01:01:09.249 INFO Starting recap_sweep re-indexing process for date: 2025-04-09 00:00:00
```

We can notice a few things:

- The re-indexing process now properly accounts for documents that belong to the previous date, in this case, 2025-04-09 00:00:00.
- The re-indexing process for the `recap_sweep` index is taking around 1 hour and 45 minutes, while the process for `recap_document_sweep` is taking around 1 hour and 27 minutes.
- The re-indexing times may vary each day since they depend on the number of documents added or updated during that day. However, these durations seem reasonable so far. As a result, the entire sending/scheduling alert process takes approximately 3 hours during the night

I also noticed a lot of noise in some related logs, which I addressed here: https://github.com/freelawproject/courtlistener/pull/5372, along with a few other small improvements.

Once that PR is merged, I’ll check the logs again. If they look good, I think we can close this one.

**Comment by mlissner** [comment]
I'm surprised by how many items are put in the sweep indexes. This is how many are modified in the last 24 hours? I would have guessed tens of thousands of cases modified and maybe 250k documents. These numbers seem high?

**Comment by albertisfu** [comment]
I reviewed the logs and confirmed that alerts triggered by the RECAP sweep index are being properly triggered every day.


> I'm surprised by how many items are put in the sweep indexes. This is how many are modified in the last 24 hours? I would have guessed tens of thousands of cases modified and maybe 250k documents. These numbers seem high?


Yes, that's correct,  these are documents added or modified in the last 24 hours, and they include the following.

(I used the cl-replica to perform the counts from the database and compare them with the numbers from Elasticsearch)

- Dockets added/modified in the last 24 hours
```
SELECT COUNT(*)
FROM search_docket
WHERE date_modified >= '2025-04-13 07:00:00+00:00'::timestamptz
  AND date_modified < '2025-04-14 07:00:00+00:00'::timestamptz;
```
**42664**


- RECAPDocuments that belong to dockets added or modified in the last 24 hours
```
SELECT COUNT(*)
FROM search_recapdocument
INNER JOIN search_docketentry ON search_recapdocument.docket_entry_id = search_docketentry.id
INNER JOIN search_docket ON search_docketentry.docket_id = search_docket.id
WHERE search_docket.date_modified >= '2025-04-13 07:00:00+00:00'::timestamptz
  AND search_docket.date_modified < '2025-04-14 07:00:00+00:00'::timestamptz;
```

This query is timing out on the replica. Could you please run it when you have a moment?
This is the query that contributes the most documents, since we need to index every RECAPDocument that belongs to a docket that was added or modified in order to match cross-object alert queries.


- RECAPDocuments added/modified in the last 24 hours
```
SELECT COUNT(*)
FROM search_recapdocument
WHERE date_modified >= '2025-04-13 07:00:00+00:00'::timestamptz
  AND date_modified < '2025-04-14 07:00:00+00:00'::timestamptz;
```
**60200**

Dockets that are parents of RECAPDocuments added/modified in the last 24 hours

```
SELECT COUNT(*)
FROM (
    SELECT DISTINCT search_docket.id
    FROM search_docket
    INNER JOIN search_docketentry ON search_docket.id = search_docketentry.docket_id
    INNER JOIN search_recapdocument ON search_docketentry.id = search_recapdocument.docket_entry_id
    WHERE search_recapdocument.date_modified >= '2025-04-13 07:00:00+00:00'::timestamptz
      AND search_recapdocument.date_modified < '2025-04-14 07:00:00+00:00'::timestamptz
) AS subquery;
```
**13119**

With the remaining query, we can compare the numbers with the ones from Elasticsearch.

**Comment by mlissner** [comment]
Here you go: 

```
courtlistener=> SELECT COUNT(*)
courtlistener-> FROM search_recapdocument
courtlistener-> INNER JOIN search_docketentry ON search_recapdocument.docket_entry_id = search_docketentry.id
courtlistener-> INNER JOIN search_docket ON search_docketentry.docket_id = search_docket.id
courtlistener-> WHERE search_docket.date_modified >= '2025-04-13 07:00:00+00:00'::timestamptz
courtlistener->   AND search_docket.date_modified < '2025-04-14 07:00:00+00:00'::timestamptz;
 count  
--------
 919697
```

So that's pretty solid confirmation that the numbers are basically right. I guess eventually we'll want to think about ways to make this faster, but for now we can close this?
